src/runtime_src/core/include/xcl_api_macros.h

```diff
@@ -18,43 +18,54 @@
 //Initialize Messages
 #include "xcl_macros.h"
 
+#define SCOPE_GUARD_MUTEX() \
+if ( sock->server_started == false ) { std::cerr<<"\n socket communication is not possible now!"; exit(0);  }\
+std::lock_guard<std::mutex> socketlk{mtx}; 
+
 
 #define AQUIRE_MUTEX() \
-mtx.lock(); \
+if ( sock->server_started == false ) { std::cerr<<"\n socket communication is not possible now!"; exit(0);  }\
+std::lock_guard<std::mutex> socketlk{mtx}; 
+
+
+// dummy Release Mutex Call.
+//#define RELEASE_MUTEX() 
+
 
-#define RELEASE_MUTEX() \
-mtx.unlock();
 
 #define RPC_PROLOGUE(func_name) \
     auto _s_inst = sock;  \
     func_name##_call c_msg; \
     func_name##_response r_msg; \
-    AQUIRE_MUTEX()
+    SCOPE_GUARD_MUTEX()
+
+
+//    AQUIRE_MUTEX() - will be deleted.
 
 #if GOOGLE_PROTOBUF_VERSION < 3006001
 // Use the deprecated 32 bit version of the size
 #define SERIALIZE_AND_SEND_MSG(func_name)                               \
     auto c_len = c_msg.ByteSize();                                      \
     buf_size = alloc_void(c_len);                                       \
     bool rv = c_msg.SerializeToArray(buf,c_len);                        \
-    if(rv == false){std::cerr<<"FATAL ERROR:protobuf SerializeToArray failed"<<std::endl;exit(1);} \
+    if(rv == false){std::cerr<<"FATAL ERROR:protobuf SerializeToArray failed for allc_void call"<<std::endl; exit(1);} \
```

Comment:
I guess the message has something to do with the real functions?
```suggestion
    if (!rv) { std::cerr << "FATAL ERROR:protobuf SerializeToArray failed for alloc_void call" << std::endl; std::exit(1); } \
```